service/storage_service.cc

```diff
@@ -4773,6 +4780,130 @@ future<> storage_service::wait_for_topology_not_busy() {
     }
 }
 
+semaphore& storage_service::get_do_sample_sstables_concurrency_limiter() {
+    return _do_sample_sstables_concurrency_limiter;
+}
+
+future<uint64_t> storage_service::estimate_total_sstable_volume(table_id t) {
+    co_return co_await seastar::map_reduce(
+        _db.local().get_token_metadata().get_host_ids(),
+        [&] (auto h) -> future<uint64_t> {
+            return ser::storage_service_rpc_verbs::send_estimate_sstable_volume(&_messaging.local(), h, t);
+        },
+        uint64_t(0),
+        std::plus<uint64_t>()
+    );
+}
+
+future<std::vector<std::byte>> storage_service::train_dict(utils::chunked_vector<bytes> sample) {
+    std::vector<std::vector<std::byte>> tmp;
+    for (const auto& s : sample) {
+        auto v = std::as_bytes(std::span(s));
+        tmp.push_back(std::vector<std::byte>(v.begin(), v.end()));
+    }
+    co_return co_await container().invoke_on(0, [tmp = std::move(tmp)] (auto& local) {
+        if (!local._train_dict) {
+            on_internal_error(slogger, "retrain_dict: _train_dict not plugged");
+        }
+        return local._train_dict(std::move(tmp));
+    });
+}
+
+future<> storage_service::publish_new_sstable_dict(table_id t_id, std::span<const std::byte> dict, service::raft_group0_client& group0_client) {
+    co_await container().invoke_on(0, coroutine::lambda([t_id, dict, &group0_client] (storage_service& local_ss) -> future<> {
+        while (true) {
+            try {
+                auto name = fmt::format("sstables/{}", t_id);
+                slogger.debug("publish_new_sstable_dict: trying to publish the dict as {}", name);
+                auto batch = service::group0_batch(co_await group0_client.start_operation(local_ss.get_abort_source()));
+                auto write_ts = batch.write_timestamp();
+                auto new_dict_ts = db_clock::now();
+                auto data = bytes(reinterpret_cast<const bytes::value_type*>(dict.data()), dict.size());
+                auto this_host_id = local_ss._db.local().get_token_metadata().get_topology().get_config().this_host_id;
+                mutation publish_new_dict = co_await local_ss._sys_ks.local().get_insert_dict_mutation(name, std::move(data), this_host_id, new_dict_ts, write_ts);
+                batch.add_mutation(std::move(publish_new_dict), "publish new SSTable compression dictionary");
+                slogger.debug("publish_new_sstable_dict: committing");
+                co_await std::move(batch).commit(group0_client, local_ss.get_abort_source(), {});
+                slogger.debug("publish_new_sstable_dict: finished");
```

Comment:
Should we have a metric when a new dict is published so we can correlate changes in compression efficiency with it?

Reply:
Maybe another metric for compressed ratio of sstables written with a dict. Since we don't care about sstables written without a dict (which AFAIK can happen when memory usage during writing exceeds a budget), for measuring the efficiency. Maybe another metric for sstable writes that had to fallback to non-dict compression.

Reply:
Note: I didn't apply this suggestion in v3. I'm not sure what they should look like, if they are the right tool at all.

For example, if the metric is global (shared by all tables), then it might stop being a good tool for measuring the efficiency as soon as more than one table exists.

Anyway, we can add more metrics anytime.